cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,209 @@
+// the destination IP address is still the VIP. Normally, there is an agent that
+// adds the vip to the local routing table, tricking the kernel in to thinking
+// it's a local IP and allowing processes doing an accept(0.0.0.0) to receive
+// the packets. Clever.
+//
+// We don't do that. Instead, we DNAT with conntrack. This is so we don't break
+// existing connections when the vip is removed. This is useful for draining
+// connections - take ourselves out of the vip, but service existing conns.
+
+const (
+	// FIXME: support multiple interfaces
+	gcpMetadataForwardedIPs = "instance/network-interfaces/0/forwarded-ips/?recursive=true"
+	// name of nat chain for iptables masquerade rules
+	gcpMasqChainName      = "gcp-vips"
+	gcpLocalMasqChainName = "gcp-vips-local"
+)
+
+type gcpHandler struct {
+	vips      []string
+	ip4tables *iptables.IPTables
+	// GCP doesn't support IPv6
+	// ip6tables *iptables.IPTables
+	stopCh chan struct{}
+}
+
+var _ handler = &gcpHandler{}
+
+func newGCPHandler() (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(5),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcpHandler{
+		ip4tables: ipt4,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: either stop the routes service, or write downfile
+func (gcp *gcpHandler) onFailure() error {
+	select {
+	case gcp.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+	return gcp.deleteRules()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (gcp *gcpHandler) onSuccess() error {
+	if err := gcp.syncRulesOnce(); err != nil {
+		return err
+	}
+
+	// bz1930457 delete stale conntrack entries to the apiserver
+	// FIXME: assume IPv4 only because GCP doesn't support IPv6
+	for _, vip := range gcp.vips {
+		// sanity check to not flush unexpected IPs
+		// provided by the metadata service
+		ip := net.ParseIP(vip)
+		if ip == nil || !ip.IsGlobalUnicast() {
+			continue
+		}
+		filter := &netlink.ConntrackFilter{}
+		filter.AddIP(netlink.ConntrackNatSrcIP, ip)
+		if _, err := netlink.ConntrackDeleteFilter(netlink.ConntrackTable, netlink.FAMILY_V4, filter); err != nil {
+			glog.V(4).Infof("Error deleting conntrack entries for %s", ip.String())
+		}
+	}
+
+	go gcp.syncRulesUntil()
+	return nil
+
+}
+
+// https://github.com/openshift/installer/blob/master/data/data/gcp/network/lb-private.tf
+// healthy_threshold   = 3
+// unhealthy_threshold = 3
+// check_interval_sec  = 2
+// timeout_sec         = 2
+
+func (gcp *gcpHandler) successThreshold() int {
+	return 1
+}
+
+func (gcp *gcpHandler) failureThreshold() int {
+	// LB = 6 seconds, plus 10 seconds for propagation
+	return 8
+}
+
+// deleteRules deletes the iptables rules added for the VIPs
+func (gcp *gcpHandler) deleteRules() error {
```

Comment:
Ideally, we always handle traffic coming from the load balancer - even if the apiserver is maybe going down, it would be good not to drop it on the floor.

The challenge will be telling the difference between traffic from the LB, and traffic from the pods. Perhaps we need a separate rule for traffic from the default interface?